### PR TITLE
feat/177 Tags Cog

### LINKED ISSUE #177: [FEAT] Tags cog
# Overview

A cog to store "tags" created by users. These tags can be referenced at a later time using `[p]tag <tag_name>`. The idea is that this would supersede most of the functionality given by the CustomCommands core cog - with the stipulation that tags can also be created by users, and that (unlike custom commands) they don't pollute the global command namespace (eg. `[p]tag foo` instead of `[p]foo`).

# Commands
```
[p]tag <tag_name>
[p]tag search <search_term>
[p]tag create <tag_name> <tag_content>
[p]tag stats <optional[member]>
[p]tag info <tag_name>
[p]tag edit <tag_name> <new_content>
[p]tag delete <tag_name>
[p]tag alias create <alias_name> <tag_name>
[p]tag alias delete <alias_name>
[p]tag claim <tag_name>
[p]tag transfer <tag_name> <member>
```

# DoD
- [x] All members have the ability to create tags
- [x] Anyone can edit a tag *that they own*
- [x] Tags can be deleted by moderators and the member who owns said tag
- [x] Tag aliases can be created by anyone
- [x] If a member leaves the server, their tags can be claimed by anyone
- [x] Tag ownership can be transferred by a moderator or the tag owner

# Stretch
- [x] List of members who can't create tags and respective moderator commands to add/remove members from said list

### PR DESCRIPTION
**Pull request by portalBlock**
# Initial Checklist
- [x] Has @tigattack / @issy been added as a reviewer?
- [x] If applicable, have the relevant project(s), milestone(s), and label(s) been applied?
- [x] If applicable, have you added details of the cog to the readme as per [README.md](https://github.com/rHomelab/LabBot-Cogs/blob/main/README.md#cog-summaries)?

# Details
**Does this resolve an issue?**
Resolves #177 

### Features / Fixes
* Implements the tags feature as specified in the above issue. 

### Breaking Changes
* Nothing should break.

## Additional
All DoD features are complete. Features to be added yet:
* Logging that respects the config options.
* Ability to configure logging with commands.
* Searching (not quite sure what @issy had in mind for this).
* Stats and info (also not entirely sure what @issy wanted for either of these). Re: stats: Every use of tags and aliases and transfer of tag ownership is logged, so there should be plenty of statistics we can pull once that's sorted.

### CODE REVIEW

File: tags/tags.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+        If the provided query is an alias, the second returned boolean will be true.
+
+        All returned values are fully hydrated.
+
+        If no aliases are found, a zero-length array is returned.
+
+        """
+        t = None  # A single tag to return
+        a = []  # All aliases for a tag
+        tp = False
```
Comment (issy) [inline tags/tags.py:249]: Some more descriptive variable names here would make things a bit more readable?

File: tags/tags.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+            }
+        }
+
+        self.config.register_guild(**default_guild_config)
+
+    @commands.guild_only()
+    @commands.group(name="tag", pass_context=True, invoke_without_command=True)
+    async def _tag(self, ctx: commands.Context, tag: str):
+
+        async def fire_tag(t) -> bool:
```
Comment (issy) [inline tags/tags.py:65]: Not a big fan of this function being nested. Maybe raise it to a class method?

File: tags/tags.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+        async with self.config.guild(ctx.guild).tags() as tags:
+            if tag in tags:
+                to = tags[tag]
+                if to["owner"] == ctx.author.id:
+                    await ctx.send("You're already that tag owner!")
+                else:
+                    curr_owner = ctx.guild.get_member(int(to["owner"]))
+                    if curr_owner is not None:
+                        await ctx.send(f"That tag's owner is still in the guild! You can see if "
+                                       f"<@{curr_owner.id}> wants to transfer it to you.")
```
[the remark this answers is not in the document]
Comment (portalBlock) [reply tags/tags.py]: Done!

File: tags/tags.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+                    else:
+                        new_owner = ctx.author.id
+                        if "transfers" not in to:
+                            to["transfers"] = []
+                        to["transfers"].append({"from": curr_owner,
```
[the remark this answers is not in the document]
Comment (portalBlock) [reply tags/tags.py]: I had that this part all kinds of messed up. Good catch! It's fixed and using IDs now (as it was supposed to before).

File: tags/tags.py (hunk 1 above)
Comment (portalBlock) [reply tags/tags.py]: Added

File: tags/tags.py (hunk 2 above)
Comment (portalBlock) [reply tags/tags.py]: Done!

**Review by issy — APPROVED** [review]
:shipit:

**Pull request merged**